crates/streamer/src/streamer/mod.rs

```diff
@@ -142,69 +155,124 @@ pub fn webm_stream(
+    ws_frame: Framed<W, ProtocolCodeC>,
+    mut chunk_receiver: mpsc::Receiver<Vec<u8>>,
     codec: Option<String>,
     mut shutdown_signal: impl Signal,
-    mut error_receiver: tokio::sync::mpsc::Receiver<UserFriendlyError>,
+    mut error_receiver: mpsc::Receiver<UserFriendlyError>,
+    mut recording_event_receiver: mpsc::Receiver<RecordingEvent>,
+    file_droped_receiver: tokio::sync::oneshot::Receiver<()>,
 ) where
     W: tokio::io::AsyncWrite + tokio::io::AsyncRead + Unpin + Send + 'static,
 {
     use futures_util::stream::StreamExt;
+
+    let ws_frame = Arc::new(Mutex::new(ws_frame));
+    let ws_frame_clone = Arc::clone(&ws_frame);
+    // Spawn a dedicated task to handle incoming messages from the client
+    // Reasoning: tokio::select! will stuck on `chunk_receiver.recv()` when there's no more data to receive
+    // This will disable the ability to receive shutdown signal
+    let handle = tokio::task::spawn(async move {
+        loop {
+            let client_message = {
+                let mut ws_frame = ws_frame.lock().await;
+                ws_frame.next().await
+            };
+
+            match client_message {
+                None => {
+                    break;
+                }
+                Some(Err(e)) => {
+                    warn!(error = %e, "error while receiving message from client");
+                    break;
+                }
+                Some(Ok(protocol::ClientMessage::Start)) => {
+                    ws_frame
+                        .lock()
+                        .await
+                        .send(protocol::ServerMessage::MetaData {
+                            codec: codec
+                                .as_ref()
+                                .and_then(|c| c.as_str().try_into().ok())
+                                .unwrap_or(protocol::Codec::Vp8),
+                        })
+                        .await?;
+                }
+                Some(Ok(protocol::ClientMessage::Pull)) => match chunk_receiver.recv().await {
+                    Some(data) => {
+                        ws_frame
+                            .lock()
+                            .await
+                            .send(protocol::ServerMessage::Chunk(&data))
+                            .await?;
+                    }
+                    None => {
+                        break;
+                    }
+                },
+            }
+        }
+
+        Ok::<_, anyhow::Error>(())
+    });
+
     let task = async move {
         info!("Starting streaming task");
-
         loop {
             tokio::select! {
                 err = error_receiver.recv() => {
                     if let Some(err) = err {
-                        ws_frame.send(protocol::ServerMessage::Error(err)).await?;
+                        let _ = ws_frame_clone.lock().await.send(protocol::ServerMessage::Error(err)).await.inspect_err(|e| {
+                            warn!(error=?e, "failed to send error message");
```

Comment:
Error messages (the thing shown by the Display implementation of error types) should be lowercased: https://github.com/Devolutions/IronRDP/blob/4b1dbaa9102c6506d8c5124d3e8a5eb0d79efd5e/STYLE.md#formatting-of-error-messages

The logging messages are not intended to be composed and may be upper cased for consistency: https://github.com/Devolutions/IronRDP/blob/4b1dbaa9102c6506d8c5124d3e8a5eb0d79efd5e/STYLE.md#logging

Reply:
Thank you! I updated them all again. I misunderstood it!